crates/aptos/src/common/types.rs

```diff
@@ -1683,62 +1739,211 @@ pub struct ArgWithTypeVec {
     /// quotes based on your shell interpreter)
     ///
     /// Example: `address:0x1 bool:true u8:0 u256:1234 "bool:[true, false]" 'address:[["0xace", "0xbee"], []]'`
-    ///
-    /// Vector is wrapped in a reusable struct for uniform CLI documentation.
     #[clap(long, multiple_values = true)]
     pub(crate) args: Vec<ArgWithType>,
 }
 
+impl TryFrom<&Vec<ArgWithTypeJSON>> for ArgWithTypeVec {
+    type Error = CliError;
+
+    fn try_from(value: &Vec<ArgWithTypeJSON>) -> Result<Self, Self::Error> {
+        let mut args = vec![];
+        for arg_json_ref in value {
+            let function_arg_type = FunctionArgType::from_str(&arg_json_ref.arg_type)?;
+            args.push(function_arg_type.parse_arg_json(&arg_json_ref.arg_value)?);
+        }
+        Ok(ArgWithTypeVec { args })
+    }
+}
+
+impl TryInto<Vec<TransactionArgument>> for ArgWithTypeVec {
+    type Error = CliError;
+
+    fn try_into(self) -> Result<Vec<TransactionArgument>, Self::Error> {
+        let mut args = vec![];
+        for arg in self.args {
+            args.push(arg.try_into()?);
+        }
+        Ok(args)
+    }
+}
+
+impl TryInto<Vec<Vec<u8>>> for ArgWithTypeVec {
+    type Error = CliError;
+
+    fn try_into(self) -> Result<Vec<Vec<u8>>, Self::Error> {
+        Ok(self
+            .args
+            .into_iter()
+            .map(|arg_with_type| arg_with_type.arg)
+            .collect())
+    }
+}
+
+impl TryInto<Vec<serde_json::Value>> for ArgWithTypeVec {
+    type Error = CliError;
+
+    fn try_into(self) -> Result<Vec<serde_json::Value>, Self::Error> {
+        let mut args = vec![];
+        for arg in self.args {
+            args.push(arg.to_json()?);
+        }
+        Ok(args)
+    }
+}
+
 /// Common options for constructing an entry function transaction payload.
 #[derive(Debug, Parser)]
 pub struct EntryFunctionArguments {
```

Comment:
I'm on the fence about this approach vs an enum where one option is EntryFunctionArguments is now and another is the JSON file option. I think it would make for cleaner downstream code. If we can make the CLI experience the same with restructuring the code this way it'd be worth doing, that way we can avoid things like `function_id.unwrap()`. Could you experiment with that?

Reply:
@banool I agree that an enum approach would be useful here, and ideally it would look something like this (for this example I'm using `ScriptFunctionArguments` since its assorted `impl` blocks are less complex):

```rust
/// Common options for constructing a script payload
#[derive(ArgEnum, Clone)]
pub enum ScriptFunctionArguments {
    CLI(ScriptFunctionArgumentsCLI),
    JSON(ScriptFunctionArgumentsJSONfile),
}

#[derive(Debug, Parser)]
pub struct ScriptFunctionArgumentsCLI {
    #[clap(flatten)]
    pub(crate) type_arg_vec: TypeArgVec,
    #[clap(flatten)]
    pub(crate) arg_vec: ArgWithTypeVec,
}

#[derive(Clone, Debug, Parser)]
pub struct ScriptFunctionArgumentsJSONfile {
    /// JSON file specifying type arguments and arguments.
    #[clap(long, parse(from_os_str))]
    pub(crate) json_file: PathBuf,
}

impl ScriptFunctionArguments {
    pub fn create_script_payload(self, bytecode: Vec<u8>) -> CliTypedResult<TransactionPayload> {
        let script_function_args: ScriptFunctionArgumentsCLI = self.try_into()?;
        Ok(TransactionPayload::Script(Script::new(
            bytecode,
            script_function_args.type_arg_vec.try_into()?,
            script_function_args.arg_vec.try_into()?,
        )))
    }
}

impl TryInto<ScriptFunctionArgumentsCLI> for ScriptFunctionArguments {
    type Error = CliError;

    fn try_into(self) -> Result<ScriptFunctionArgumentsCLI, Self::Error> {
        match self {
            ScriptFunctionArguments::CLI(args) => Ok(args),
            ScriptFunctionArguments::JSON(path_wrapper) => {
                Ok(
                    parse_json_file::<ScriptFunctionArgumentsJSON>(&path_wrapper.json_file)?
                        .try_into()?,
                )
            },
        }
    }
}

#[derive(Deserialize)]
/// JSON file format for script function arguments.
struct ScriptFunctionArgumentsJSON {
    type_args: Vec<String>,
    args: Vec<ArgWithTypeJSON>,
}

impl TryInto<ScriptFunctionArgumentsCLI> for ScriptFunctionArgumentsJSON {
    type Error = CliError;

    fn try_into(self) -> Result<ScriptFunctionArgumentsCLI, Self::Error> {
        Ok(ScriptFunctionArgumentsCLI {
            type_arg_vec: TypeArgVec::try_from(&self.type_args)?,
            arg_vec: ArgWithTypeVec::try_from(&self.args)?,
        })
    }
}
```

The catch, however, is that the `enum` approach only works for unit variants:

```bash
error: `#[derive(ValueEnum)]` only supports unit variants. Non-unit variants must be skipped
    --> crates/aptos/src/common/types.rs:1887:5
     |
1887 |     CLI(ScriptFunctionArgumentsCLI),
     |     ^^^

```

It does appear that the kind of `enum` support we'd both like is actually an open feature request in `clap`: https://github.com/clap-rs/clap/issues/2621#issuecomment-1058333744, but until that is made available I think the most straightforward is a `derive`-restricted command of the form: https://github.com/aptos-labs/aptos-core/blob/6aa74479f6b3f2b44f06cc0aea7efbd9acff1a51/crates/aptos/src/common/types.rs#L1801-L1818

Here I've included your suggestions per https://github.com/aptos-labs/aptos-core/pull/8346#discussion_r1205989174 to create a situation that wherever entry function arguments are accepted, they can be passed either via the CLI interpreter or via an input file.

Your thoughts?